wandb/util.py

```diff
@@ -546,8 +546,13 @@ def request_with_retry(func, *args, **kwargs):
             response.raise_for_status()
             return response
         except (requests.exceptions.ConnectionError,
-                requests.exceptions.HTTPError,  # XXX 500s aren't retryable
+                requests.exceptions.HTTPError,
                 requests.exceptions.Timeout) as e:
+            if isinstance(e, requests.exceptions.HTTPError):
+                if e.response.status_code in {400, 403, 404, 409, 500}:
```

Comment:
We always want to retry 5xx errors in case of an outage.  We should make the backend return a 4xx error if it can't handle a request.

Reply:
I think you’re thinking of 501, 503, and others. Straight 500 means eg. a bug on the server and should never be retried. So yeah, I’d agree that returning 4xx for invalid JSON might make sense but a server bug wouldn’t count as one of those.

Reply:
5xx can mean MySQL or PubSub went down, it essentially means "Unknown error".  For instance when we deploy we throw some 5xx's because of lost MySQL connections.  501-503 are mostly some kind of timeout either at the gateway or the backend.  4xx are "Known errors" which we can safely never retry.  If we start getting a bunch of 5xx's on the backend we're paged and can fix the underlying issue allowing the clients to keep retrying.

Reply:
I guess I’m saying we should distinguish between temporary errors and permanent errors. It could be communicated through the error code or some other way. 500 is the default when there’s an exception on the server.  The web app currently does a ton of pointless retries when the back end throws any kind if exception. If that is also what happens when MySQL is failing, maybe it’d be good to return some kind of flag or just use a different error code. I’d guess MySQL goes down way less often than the back end fails some other way. It probably doesn’t make sense to retry on 500s, but we should indicate some other way that the client should retry when MySQL fails.

Reply:
Huh, it looks like I was wrong about 501. It’s “not implemented”. But 503 is “service unavailable,” which seems to me like a better fit for MySQL being down than 500. It may also be used for load balancers and things, but that doesn’t mean it shouldn’t be used for other infrastructure stuff